### ISSUE TITLE
intrinsics runtime detection prevents release

### ISSUE THREAD

**Issue by patricklodder** [dogecoin/dogecoin#3134]
Reason for flagging this up
----------------------------

This morning, I received a question about eecffe5 and whether we have patched that, because it's the cause of [CVE-2018-17144](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2018-17144). Fortunately, we have, with 52965fb. 

This did however cause some stress because we currently have a broken 1.14.7-dev branch due to e78cfc6 from #3091 breaking intrinsics out of experimental, breaking compilation on for example M1 macOS because the detection needs to be customized for Apple OSs. This would make it hard to rapidly release 1.14.7 if we have an urgent need, because we do not distribute native binaries for M1 so we'd need to start creating patch releases like `1.14.6.1` to support all previously working self-compile processes. Not having backward compatibility on as many platforms as possible for critical bugs is 

Example issue reproduction
-------------------

To reproduce this on an M1 by building a small target:
```sh
brew install autoconf automake libtool openssl boost
./autogen.sh
./configure --with-gui=no --disable-wallet --with-boost=/opt/homebrew/Cellar/boost/1.80.0 \
   CPPFLAGS="-I/opt/homebrew/Cellar/boost/1.80.0/include"
make -j8 -C src bench
```

will result in:

```
crypto/sha1.cpp:373:19: error: use of undeclared identifier 'AT_HWCAP'
    if (getauxval(AT_HWCAP) && HWCAP_SHA1)
                  ^
crypto/sha1.cpp:373:32: error: use of undeclared identifier 'HWCAP_SHA1'
    if (getauxval(AT_HWCAP) && HWCAP_SHA1)
                               ^
2 errors generated.
```

Proposed solution
-------------------

I have been looking at just fixing this non-intrusively, but even though [introducing Apple-specific capability detection](https://developer.apple.com/documentation/kernel/1387446-sysctlbyname/determining_instruction_set_characteristics) is not all that hard, it actually is hard to be sure that nothing else breaks on different platforms that I either do not have easy access to, or forget about. So this is a risky approach.

To not be in this situation for much longer (and not have a heart attack every time something that may become urgent pops up) I'm going to propose a couple of PRs today/tomorrow:

- [ ] Revert eecffe5 and for the time being freeze any intrinsics work, until we have:
- [ ] Introduced and enforced hard checks on experimental features so that we do not accidentally break experimental features out in the future (won't take ages, I'm working on this right this moment), after which:
- [ ] @edtubbs can simply re-propose his work and we can test it more thoroughly, while keeping also the runtime capability checks guarded under experimental feature flags until we are sure about compatibility and introduce means to actually turn it off if we need to.

I think that by doing it like this, we will have a not-broken development branch, set ourselves up to better prevent future human error (or at least make it easier to detect) regarding experimental features and can then resume the work on intrinsics and runtime detection, because this actually is important work.

Thoughts and feedback welcome - I am regardless of that preparing 2 PRs as proposed above in the meantime

**Comment by edtubbs** [comment]
Hey @patricklodder, thanks for tracking this down!  Unfortunately the continuous integration didn't catch this, perhaps if we had a native Mac build or more time in PR.  Either way, it should be possible to add the logic for Apple to Initialize_transform_ptr(void) and setup the function pointer correctly.  If you agree with that approach, I can proceed with further investigation.

**Comment by edtubbs** [comment]
> Yes, that can work for Apple, but I don't know about other non-gnu platforms. That's why I suggest we revert, add a tiny framework (I'm past 50% in so I'll PR it latest tomorrow early hours) and then we should be able to take a little more time on review. I'm basically trying to get it out of showstopper state.

Ok cool, agreed.  I'll look forward to that PR and we'll go from there.  Thanks!

**Comment by patricklodder** [comment]
@edtubbs the 2nd PR is #3136 - please let us know if you think that is workable.

**Comment by edtubbs** [comment]
> @edtubbs the 2nd PR is #3136 - please let us know if you think that is workable.

Yes, and we can continue to work together on the approach for runtime checks.
